[JIT] Add support for named_modules()

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#29495 [JIT] Add support for named_modules()**


Fix for https://github.com/pytorch/pytorch/issues/28998

Differential Revision: [D18412561](https://our.internmc.facebook.com/intern/diff/D18412561)

Commit:
add support for _modules, reducing special casing of nn.Sequential

[ghstack-poisoned]

Commit:
Update on "add support for _modules, reducing special casing of nn.Sequential"


This PR adds support for `_modules`, making it so we no longer need to special case support for `nn.Sequential`. I was getting internal errors around the previous approach using `self.define()`, so i am adding this PR as part of the stack.

Fix for https://github.com/pytorch/pytorch/issues/28998

[ghstack-poisoned]

driazati commented on torch/csrc/jit/script/python_sugared_value.h:
```diff
@@ -102,6 +102,70 @@ struct VISIBILITY_HIDDEN ConstantParameterList : public SugaredValue {
+      iterator->addChild(loc, m, modules_);
+      return std::make_shared<ModuleDictMethod>(iterator, "items");
+    } else {
+      throw ErrorReport(loc) << "_Modules only supports keys, values, or items";
```
`_Modules` -> `_modules`

zdevito reviewed (dismissed):
Binding a private python member into TorchScript seems like the wrong approach. Shouldn't we implement the standard nn.Module API for this stuff? That would be the `children`, `named_children`, `modules()` and `named_modules` methods. Having `_modules` without `named_children` is going to promote the use of a private API.

Commit:
Update on "add support for _modules, reducing special casing of nn.Sequential"


This PR adds support for `_modules`, making it so we no longer need to special case support for `nn.Sequential`. I was getting internal errors around the previous approach using `self.define()`, so i am adding this PR as part of the stack.

Fix for https://github.com/pytorch/pytorch/issues/28998

Differential Revision: [D18412561](https://our.internmc.facebook.com/intern/diff/D18412561)

[ghstack-poisoned]

Commit:
Update on "add support for _modules, reducing special casing of nn.Sequential"


This PR adds support for `_modules`, making it so we no longer need to special case support for `nn.Sequential`. I was getting internal errors around the previous approach using `self.define()`, so i am adding this PR as part of the stack.

Fix for https://github.com/pytorch/pytorch/issues/28998

Differential Revision: [D18412561](https://our.internmc.facebook.com/intern/diff/D18412561)

[ghstack-poisoned]

Commit:
Update on "add support for _modules, reducing special casing of nn.Sequential"


This PR adds support for `_modules`, making it so we no longer need to special case support for `nn.Sequential`. I was getting internal errors around the previous approach using `self.define()`, so i am adding this PR as part of the stack.

Fix for https://github.com/pytorch/pytorch/issues/28998

Differential Revision: [D18412561](https://our.internmc.facebook.com/intern/diff/D18412561)

[ghstack-poisoned]

Commit:
Update on "add support for _modules, reducing special casing of nn.Sequential"


This PR adds support for `_modules`, making it so we no longer need to special case support for `nn.Sequential`. I was getting internal errors around the previous approach using `self.define()`, so i am adding this PR as part of the stack.

Fix for https://github.com/pytorch/pytorch/issues/28998

Differential Revision: [D18412561](https://our.internmc.facebook.com/intern/diff/D18412561)

[ghstack-poisoned]

eellison commented:
This doesn't do the checking for uniqueness, but when we recursively script we create a new module on each compilation (don't preserve aliasing), so i don't think that's actually a problem.

Commit:
Update on "[JIT] Add support for named_modules()"



Fix for https://github.com/pytorch/pytorch/issues/28998

Differential Revision: [D18412561](https://our.internmc.facebook.com/intern/diff/D18412561)

[ghstack-poisoned]

driazati commented on torch/csrc/jit/script/python_sugared_value.h:
```diff
@@ -144,40 +175,71 @@ struct VISIBILITY_HIDDEN ModuleValue : public SugaredValue {
-    }
-    return iterable_;
+      Function& m,
+      const std::string& field) override {
```
Can you move the implementation to the `.cpp` file

driazati commented on torch/csrc/jit/script/python_sugared_value.cpp:
```diff
@@ -316,24 +334,16 @@ std::shared_ptr<SugaredValue> ModuleValue::attr(
-      }
-      return std::make_shared<ModuleDictMethod>(
-          desugarModuleContainer(get_keys, get_values, loc, m), field);
+    if (field == "items" || field == "keys" || field != "values") {
```
This should be `field == "values"` right

driazati reviewed (approved):
Can you make a follow up issue or add to this PR stuff like `named_children`?

eellison commented:
> Can you make a follow up issue or add to this PR stuff like `named_children`?

i'm going to do the other apis as a follow up sometime this week.

Commit:
Update on "[JIT] Add support for named_modules()"



Fix for https://github.com/pytorch/pytorch/issues/28998

Differential Revision: [D18412561](https://our.internmc.facebook.com/intern/diff/D18412561)

[ghstack-poisoned]

eellison commented:
@chenbohua3  thx !
